crates/ark/src/reticulate.rs

```diff
@@ -0,0 +1,145 @@
+
+lazy_static! {
+    static ref RETICULATE_COMM_ID: Mutex<Option<String>> = Mutex::new(None);
+}
+
+pub struct ReticulateService {
+    comm: CommSocket,
+    comm_manager_tx: Sender<CommManagerEvent>,
+}
+
+impl ReticulateService {
+    fn start(comm_id: String, comm_manager_tx: Sender<CommManagerEvent>) -> anyhow::Result<String> {
+        let comm = CommSocket::new(
+            CommInitiator::BackEnd,
+            comm_id.clone(),
+            String::from("positron.reticulate"),
+        );
+
+        let service = Self {
+            comm,
+            comm_manager_tx,
+        };
+
+        let event = CommManagerEvent::Opened(service.comm.clone(), serde_json::Value::Null);
+        unwrap!(service.comm_manager_tx.send(event), Err(e) => {
+            log::error!("Reticulate: Could not open comm. Error {e}");
+        });
+
+        spawn!(format!("ark-reticulate-{}", comm_id), move || {
+            unwrap!(service.handle_messages(), Err(err) => {
+                log::error!("Connection Pane: Error while handling messages: {err:?}");
+            });
+        });
+
+        Ok(comm_id)
+    }
+
+    fn handle_messages(&self) -> Result<(), anyhow::Error> {
+        loop {
+            let msg = unwrap!(self.comm.incoming_rx.recv(), Err(err) => {
+                log::error!("Reticulate: Error while receiving message from frontend: {err:?}");
+                break;
+            });
+
+            if let CommMsg::Close = msg {
+                self.comm.outgoing_tx.send(CommMsg::Close).unwrap();
+                break;
+            }
+
+            // Forward data msgs to the frontend
+            if let CommMsg::Data(_) = msg {
+                self.comm.outgoing_tx.send(msg)?;
+                continue;
+            }
+        }
+
+        // before finalizing the thread we make sure to send a close message to the front end
+        if let Err(err) = self.comm.outgoing_tx.send(CommMsg::Close) {
+            log::error!("Reticulate: Error while sending comm_close to front end: {err:?}");
+        }
+
+        let mut comm_id_guard = unwrap!(
+            RETICULATE_COMM_ID.try_lock(),
+            Err(e) => {
+                return Err(anyhow!("Could not access comm_id. Error {}", e));
+            }
+        );
+        log::info!(
+            "Reticulate Thread closing {}",
+            (*comm_id_guard).clone().unwrap()
+        );
+
+        *comm_id_guard = None;
+
+        Ok(())
+    }
+}
+
+// Creates a client instance reticulate can use to communicate with the front-end.
+// We should aim at having at most **1** client per R session.
+// Further actions that reticulate can ask the front-end can be requested through
+// the comm_id that is returned by this function.
+#[harp::register]
+pub unsafe extern "C" fn ps_reticulate_open(input: SEXP) -> Result<SEXP, anyhow::Error> {
+    let main = RMain::get();
+
+    if !RMain::initialized() {
+        return Ok(R_NilValue);
+    }
+
+    let input = RObject::try_from(input)?;
```

Comment:
We're converging towards this style which makes for clearer code thanks to the type annotation:

```suggestion
    let input: RObject = input.try_into()?;
```

Reply:
Also for consistency with the line below.